### ISSUE TITLE
Emoji Support

### ISSUE THREAD

**Issue by loesler** [ilosuna/mylittleforum#642]
One reason to switch to `utf8mb4` was to support emoji, cf. [Umbau der Datenbank für Unicode-Emojis](https://mylittleforum.net/forum/index.php?id=11777). However, up to know, smileys are represented by images and one has to know the emoji-code. 
For that reason, I created a [test page](https://mylittleforum.net/test/emoji/) that contains a proposal for adding unicode emoji to mlf. Currently, the [full list of emoji](https://unicode.org/emoji/charts/emoji-list.html) is added regardless of whether an emoji-icon exists. 

1) Is such an implementation /better/ than the current one?
2) How to handle non existing emoji-icons?
3) How to handle IE11?
4) Any other suggestions?

Please check out this [Emoji-Testpage](https://mylittleforum.net/test/emoji/).

**Comment by Helmut01** [comment]
Hi Micha,

just observations in my forum. Whether emojis are displayed in their doubtful colored beauty seemingly depends on the browser. Try my forum in [mix view](https://forum.bebac.at/mix.php) and esp. the first three posts. Checked in the latest browser version on Windows:

1. Firefox, SeaMonkey, Tor: Emojis – as expected – in color.
2. Chrome, Opera, Edge, IE11: Emojis in the color of the text, the pushpin in black.

Maybe I screwed up somewhere.

**Comment by loesler** [comment]
Hi Helmut,

in Opera, the emojis are colored, cf.
![emoji](https://user-images.githubusercontent.com/3595603/189491877-72db58f3-52ee-4adf-900c-f55e5bc77568.png)


However, the color is not the main problem. At least in Opera, some emojis seem to be undefined. These emojis are shown as blue rectangle in the following screenshot. How to handle these emojis? 

![emoji_picker](https://user-images.githubusercontent.com/3595603/189492006-f70498f9-0883-4f1f-a824-d52413e05ffc.png)

/Micha

**Comment by Helmut01** [comment]
Hi Micha,

not on mine (v90.0.4480.84 on Windows):
![Opera](https://user-images.githubusercontent.com/19939535/189517968-52c30190-dc7e-467a-aaac-3f8c22fdd23c.png)

I’m afraid we can do nothing with the “missing” ones. That’s why I stated a warning in my forum’s [Instructions](https://forum.bebac.at/Instructions.htm#UTF8Warning). BTW, depending on the browser the capsule in the example looks like this
![image](https://user-images.githubusercontent.com/19939535/189518843-21d6757f-6d4e-4178-8df1-fe852b68c309.png) (19×17px) or 
![image](https://user-images.githubusercontent.com/19939535/189518880-22666348-aa3a-4951-be41-10bb61884155.png) (21×14px)
If we want to replace the GIF-smilies with their emoji-counterparts we would have to check a lot of browsers (-versions) on many OSs (Windows, Linuxes, OSX, Android, …). Do we really want to go there?

**Comment by auge8472** [comment]
The placeholder rectangles are unicode code points for chars/emojis, the browser does not find the char/emoji for. Some browsers in some operating systems will find a font with a char for this code point, some in some combination not. See therefore my following screenshot from the test page. Firefox in Ubuntu 20.04 with Mate desktop. There are emojis from at least two fonts (one coloured, one with only a contour of the emoji) and several chars with the placeholder for not found chars.

![mlf2-emoji-list](https://user-images.githubusercontent.com/1734660/189526811-14b38e6f-903e-4b69-a078-5ee96ceecac8.png)

To make the confusion complete I show now a screenshot of the same section of the same function in the SelfHTML-forum, made with the same combination of browser and OS. All chars are displayed, no placeholders, no different font.

![mlf2-emoji-list-selfforum](https://user-images.githubusercontent.com/1734660/189527381-58041762-74e7-43d0-b24d-b6f417a2a91a.png)

Even, when I add the font-family-rule from the SelfHTML-forum to [my own version of the MLF-test-page](https://mylittleforum.net/test/emoji/index-auge.html), the emoji-list remains incomplete. :-/

**Comment by loesler** [comment]
Hello,

I uploaded a font and - if the font is loaded - it works. However, this is not my preferred *smart* solution. If this is the only *robust* way, we should revert to the normal PNG files...

/Micha

**Comment by auge8472** [comment]
For comparision: The emoji-selection-bar in Windows 10 at work with Firefox (as far as I know without any strange configuration):

![mlf-emojji-selection-windows-10](https://user-images.githubusercontent.com/1734660/189676711-9ed4eec6-d08f-428c-bba4-b294d7bca581.png)

Same view in Chrome (similar to your screenshots).

> I uploaded a font and - if the font is loaded - it works. However, this is not my preferred smart solution. If this is the only robust way, we should revert to the normal PNG files...

We need further research. Other services got it running.

**Comment by loesler** [comment]
> For comparision: The emoji-selection-bar in Windows 10 at work with Firefox (as far as I know without any strange configuration)

In your screenshot, the external font is not loaded. It is easy to verify, that different fonts are used in your screenshot to display emojis. If the external font is available/downloaded, the emojis look more unified, cf.

![emoji](https://user-images.githubusercontent.com/3595603/189695785-f1e12d94-94d1-4eaf-a00f-aead5329da7d.png)

Maybe, you have to open the box several times. I checked the result in Edge, Opera and Firefox on two different computers (both with win10).

> We need further research. Other services got it running.
I believe, these services make use of external fonts, too.

**Comment by auge8472** [comment]
Sorry for causing misunderstanding. My comparision (Linux and Windows) affected my own test page with the same font-falily-settings as in the SelfHTML.org-forum. I expected a different outcome because of different installed fonts on my personal computer (Linux) and the machine for my job (Windows). But as one can see, there are the same chars missing and the same chars in a different font in the screenshots from both machines.

In your test page with the web font the missing emojis are still missing but the others seems to be using all the same font.

**Comment by auge8472** [comment]
CoVid time (more or less over) brings boredom.

I compared the solution in the SelfHTML forum again with our attempts. I took over the font-family-settings for the postings and the Emoji-selection-box, no success. I checked, if my persomal computer had installed a proper font and installed one free emoji font (Noto Emoji Regular) on my system, no success.

The only solution was to use a web font with `@font-face`. Because the only for free downloadable file type of the font was TTF with a file size of around 24MB (what is out of question) I converted the font to WOFF (2.3MB) and WOFF2 (5.4MB). With that step I can see now all emojis. I wonder, why the WOFF2 file is so much larger than the WOFF file because the compression algorithm of WOFF2 should be so much better than the one od WOFF.

The font files are uploaded to the mylittleforum.net-webspace. And yes, even 2.3 MB is far from anything like My**Little**Font. 😉

---

A second thing is the emoji-selection-box in itself. I am working on an alternative construction. We will see, if I'll get a working result.

**Comment by loesler** [comment]
FYI: SELFHTML uses a services like https://missive.github.io/emoji-mart/

**Comment by auge8472** [comment]
I will ask for support in the SelfHTML-forum.

**Comment by auge8472** [comment]
> FYI: SELFHTML uses a service ...

Yes, [I think](https://forum.selfhtml.org/self/2022/oct/14/emoji-fonts-hier-im-forum/1802883#m1802883) they use the external service named "Emoji Mart".

While writing a feature list for MLF2 I came across the emoji issue and the necessitiy of a dedicated font again. I remembered the upgrades of messenger software, namely Text Secure/Signal, where the installation of a new version of such dedicated fonts was mentioned every when and then. So it seems to me, that even big players are not willing to count on the OS and the fonts they provide.

In general we can assume, that there are forums where emojis are very welcome and other installation and others where they do not matter. This speaks for being an optional feature. In general it would be possible to install an emoji font for the forum as the forum operator and to choose the font in the admin panel afterwards i.e. on the smilies management page. But such a font can be incomplete in regard of the number of emojis by itself. To prevent showing the user a selection box with broken emojis, it would be nice to have an automatism to detect missing emojis and to sort them out.

Currently I've no plausible idea how to handle this immature idea.
